internal/plugin/grpc_provider.go

```diff
@@ -94,53 +97,44 @@ type GRPCProvider struct {
 	// to use as the parent context for gRPC API calls.
 	ctx context.Context
 
-	mu sync.Mutex
-	// schema stores the schema for this provider. This is used to properly
-	// serialize the requests for schemas.
-	schema providers.GetProviderSchemaResponse
+	// SchemaCache stores the schema for this provider. This is used to properly
+	// serialize the requests for schemas.  This is shared between instances
+	// of the provider.
+	SchemaCache      func(func() providers.GetProviderSchemaResponse) providers.GetProviderSchemaResponse
```

Comment:
I think the func(func T) T pattern took a while for me to put together in my head what it was doing and I think its great to do a memoizing/caching approach this way.

However I actually think it would be a little bit nicer to provider a small struct that wrapped it so its a tiny bit clearer, especially at execution time. I would expect to call a method called something like `providerCache.GetOrFetch()` maybe? (Naming is hard!)

Reply:
Pushed a change to replace this with the alias